src/colvargrid_integrate.cpp

```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
+    get_grad(g000, ix); // (0,0,0)
+    ix[2] = ix0[2] + 1;
+    wrap_detect_edge(ix);
+    get_grad(g001, ix); // (0,0,1)
+
+    ix[1] = ix0[1] + 1;
+    wrap_detect_edge(ix);
+    get_grad(g011, ix); // (0,1,1)
+
+    ix[2] = ix0[2];
+    wrap_detect_edge(ix);
+    get_grad(g010, ix); // (0,1,0)
 
+    ix[0] = ix0[0] + 1;
+    ix[1] = ix0[1];
+    ix[2] = ix0[2];
+    wrap_detect_edge(ix);
+    get_grad(g100, ix); // (1,0,0)
+
+    ix[2] = ix0[2] + 1;
+    wrap_detect_edge(ix);
+    get_grad(g101, ix); // (1,0,1)
+
+    ix[1] = ix0[1] + 1;
+    wrap_detect_edge(ix);
+    get_grad(g111, ix); // (1,1,1)
+
+    ix[2] = ix0[2];
+    wrap_detect_edge(ix);
+    get_grad(g110, ix); // (1,1,0)
     divergence[linear_index] =
-     ((gc[3*4]-gc[0] + gc[3*5]-gc[3*1] + gc[3*6]-gc[3*2] + gc[3*7]-gc[3*3])
-      / widths[0]
-    + (gc[3*2+1]-gc[0+1] + gc[3*3+1]-gc[3*1+1] + gc[3*6+1]-gc[3*4+1] + gc[3*7+1]-gc[3*5+1])
-      / widths[1]
-    + (gc[3*1+2]-gc[0+2] + gc[3*3+2]-gc[3*2+2] + gc[3*5+2]-gc[3*4+2] + gc[3*7+2]-gc[3*6+2])
-      / widths[2]) * 0.25;
+        ((g100[0] - g000[0] + g101[0] - g001[0] + g110[0] - g010[0] + g111[0] - g011[0]) /
+             widths[0] +
+         (g010[1] - g000[1] + g011[1] - g001[1] + g110[1] - g100[1] + g111[1] - g101[1]) /
+             widths[1] +
+         (g001[2] - g000[2] + g011[2] - g010[2] + g101[2] - g100[2] + g111[2] - g110[2]) /
+             widths[2]) *
+        0.25;
+    cvm::real supplement = 0;
+    cvm::real supplement_coefficient = 1;
+
+    ix = ix0;
+    if ((ix0[0] == 0 || ix0[0] == computation_nx[0] - 1) && !periodic[0]) {
+      cvm::real coefficient = (ix[0] == 0) ? 1.0 : -1.0;
+      supplement_coefficient *= 0.5;
+      supplement = coefficient *
+                   (g100[0] + g000[0] + g101[0] + g001[0] + g110[0] + g010[0] + g111[0] + g011[0]) *
+                   0.125 * 2 * widths[0];
+    }
+    if ((ix0[1] == 0 || ix0[1] == computation_nx[1] - 1) && !periodic[1]) {
+      cvm::real coefficient = (ix[1] == 0) ? 1.0 : -1.0;
+      supplement_coefficient *= 0.5;
+      supplement +=
+          coefficient *
+          (g010[1] + g000[1] + g011[1] + g001[1] + g110[1] + g100[1] + g111[1] + g101[1]) *
+          widths[1] * 2 * 0.125;
+    }
+    if ((ix0[2] == 0 || ix0[2] == computation_nx[2] - 1) && !periodic[2]) {
+      cvm::real coefficient = (ix[2] == 0) ? 1.0 : -1.0;
+      supplement_coefficient *= 0.5;
+      supplement +=
+          coefficient *
+          (g001[2] - g000[2] + g011[2] - g010[2] + g101[2] - g100[2] + g111[2] - g110[2]) * 0.125 *
+          2 * widths[2];
+    }
+    divergence[linear_index] += supplement;
+    divergence[linear_index] *= supplement_coefficient;
   }
 }
 
 
+void colvargrid_integrate::prepare_divergence_stencils()
+{
+  surrounding_points_relative_positions.clear();
+  size_t n_combinations = pow(2, nd);
+  for (size_t i = 0; i < n_combinations; i++) {
+    surrounding_points_relative_positions.push_back(convert_base_two(i, nd));
```

Comment:
I feel that the calls to `convert_base_two` is unnecessary because an integer has already all its bits itself. `convert_base_two` simply checks the bits of `i` and stores them into a vector of `nd` size. But maybe I am wrong, and I am not sure how to optimize this.